### PR TITLE
Configuration using hydra

### LINKED ISSUE #19: configure models with hydra

### PR DESCRIPTION
**Pull request by alansaul**
Allowing `hydra` to be used for executing experiments. This allows easy overrides of model or acquisition features, and we can also later set up multi-run experiments. Needs a few more tests, but does introduce a smoke test for all the current models.

Example use:
`poetry run python3 scripts/run_experiment_hydra.py acquisition=ei ++acquisition.builder.batch_size=1 model=deep_gp`

Resolves #19

### CODE REVIEW

File: experiment/conf/acquisition/continuous_ts.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+defaults:
+  - base
+
+key: "continuous"
```
Comment (hstojic) [inline experiment/conf/acquisition/continuous_ts.yaml:4]:
```suggestion
key: "continuous_ts"
```

File: experiment/conf/model/deep_ensemble.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+builder:
+  _target_: experiment.models.DeepEnsembleBuilder
+
+  ensemble_size: 5
+  num_hidden_layers: 3
+  units: 25
+  activation: "selu"
+
+  bootstrap: true
+  diversify: false
```
Comment (hstojic) [inline experiment/conf/model/deep_ensemble.yaml:12]:
```suggestion
  diversify: true
```

File: experiment/conf/model/gpr.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+key: "gpr"
+
+builder:
+  _target_: experiment.models.GPRBuilder
+
+  trainable_likelihood: false
+  likelihood_variance: null
```
Comment (hstojic) [inline experiment/conf/model/gpr.yaml:7]: for deterministic objectives we probably want to set this to 1e-5 by default

File: experiment/experiment.py [hunk 4]
```diff
@@ -11,83 +11,166 @@
...
-    )
-    pd.DataFrame(result_observations).to_csv(
-        f"results/{function_key}/{acq_key}_{model_key}_{seed}_observations.csv"
-    )
 
-    print(
-        f"{acq_key} {model_key} {function_key} {seed}\n"
-        f"min observation: {result_observations[result_arg_min_idx, :]}\n"
-        f"regret: {result_observations[result_arg_min_idx, :] - F_MINIMA}\n"
+def acquisition_rule_from_config(acq_cfg: DictConfig) -> AcquisitionRule:
```
Comment (hstojic) [inline experiment/experiment.py:35]: this seems like it should go to `acquisition.py`

File: experiment/experiment.py [hunk 5]
```diff
@@ -11,83 +11,166 @@
...
+    """
+    Given a OmegaConf based configuration, instantiate an acquisition rule instance
+    :param acq_cfg: configuration specific to the acqusition rule
+    :return: a new acqusition rule instance
+    """
+    return call(acq_cfg.builder)
+
+
+@dataclasses.dataclass
+class ExperimentRunner:
```
Comment (hstojic) [inline experiment/experiment.py:45]: I have updated the runner quite a bit, we can address this after my PR is merged in

File: experiment/experiment.py [hunk 6]
```diff
@@ -11,83 +11,166 @@
...
+
+        print(
+            f"{self.acq_key} {self.model_key} {problem_name} {self.seed}\n"
+            f"min observation: {result_observations[result_arg_min_idx, :]}\n"
+            f"regret: {result_observations[result_arg_min_idx, :] - self.problem.minimum}\n"
+        )
+
+
+@hydra.main(version_base=None, config_path="conf", config_name="config")
+def run_experiment_hydra(
```
Comment (hstojic) [inline experiment/experiment.py:143]: this seems to me like it should go to the file in scripts, no?

File: experiment/models.py [hunk 7]
```diff
@@ -40,142 +42,204 @@
...
+                return lr
+
+        fit_args = {
+            "batch_size": self.batch_size,
+            "epochs": self.epochs,
+            "callbacks": [tf.keras.callbacks.LearningRateScheduler(scheduler)],
+            "verbose": 0,
+            "shuffle": False,
+        }
+        optimizer = KerasOptimizer(tf.optimizers.Adam(self.adam_lr), fit_args)
```
Comment (hstojic) [inline experiment/models.py:97]: I think we probably we want to have configs for optimizers and callbacks as well, but we can postpone that to another PR

File: experiment/experiment.py [hunk 8]
```diff
@@ -11,83 +11,166 @@
...
+    Example usage:
+        poetry run python3 scripts/run_experiment_hydra.py acquisition=ei \
+            ++acquisition.builder.batch_size=1 model=gpr
+
+    :param cfg: OmegaConf based configuration file, automatically built from command line arguments
+    :return: Optimisation result from running BO loop with given configuration
+    """
+
+    problem = problem_dict[cfg.function_key]
+    model_builder = instantiate(cfg.model.builder)
```
Comment (hstojic) [inline experiment/experiment.py:160]: this is neat!

**Review by hstojic — COMMENTED** [review]
I did a pass - overall, looks great, I think in comparison to previous config systems I have seen this feels much more lightweight!

I have changed the ExperimentRunner quite a bit in the meantime, sorry ;/

File: experiment/experiment.py (hunk 5 above)
Comment (alansaul) [reply experiment/experiment.py:45]: Yep no worries, will handle conflicts when your PR is merged; thought I would just go some way to aligning them early so it was clear to see how they would match up.

File: experiment/experiment.py (hunk 6 above)
Comment (alansaul) [reply experiment/experiment.py:143]: Yeah, gradually as I refactored the script file was whittled down to nothing, think I agree that the experiment runner doesn't need to know its being instantiated via hydra, which is nice. I'll move it.

File: .github/workflows/quality-checks.yaml [hunk 9]
```diff
@@ -27,7 +27,7 @@ jobs:
         python-version: '3.10'
     - uses: Gr1N/setup-poetry@v7
       with:
-        poetry-version: 1.1.14
+        poetry-version: 1.4.0
```
Comment (alansaul) [inline .github/workflows/quality-checks.yaml:30]: Note I've bumped the poetry version. If you're not happy with this (or want me to do it in a separate PR) please say and I'll go back to the old poetry lock file.

File: experiment/experiment.py [hunk 10]
```diff
@@ -95,9 +92,10 @@ def run(self) -> None:
             print(f"Experiment {self.experiment_name}/{self.step}: Ask complete")
 
             if self.reinitialise_models:
-                self.optimiser.models = self._initialise_models()
+                self.optimiser.models = self._initialise_models(self.initial_data)
```
Comment (alansaul) [inline experiment/experiment.py:95]: Note that here I am not changing the behaviour as it doesn't belong in this PR, but I think this should actually have the most up to date data provided.

File: experiment/experiment.py [hunk 11]
```diff
@@ -112,49 +110,47 @@ def _init(self) -> None:
...
         typically we want to do that, reusing the parameters from previous iteration leads to worse
         performance.
         """
         print(f"Experiment {self.experiment_name}/{self.step}: Compiling the models")
+        return {
+            tag: self.model_builder.create(data=data, search_space=self.problem.search_space)
+            for tag, data in data.items()
+        }
 
+    def collect_initial_data(self) -> Mapping[Tag, Dataset]:
```
Comment (alansaul) [inline experiment/experiment.py:139]: Better name? suggestions?

File: experiment/problems.py [hunk 12]
```diff
@@ -23,7 +23,7 @@
     Shekel4,
 )
 
-problem_factory = {
+problem_dict = {
```
Comment (alansaul) [inline experiment/problems.py:26]: Not technically a factory as you can only get a single instance, hence the name.

File: pyproject.toml [hunk 13]
```diff
@@ -90,7 +90,8 @@ warn_redundant_casts = true
 warn_unused_ignores = true
 strict_equality = true
 show_error_codes = true
+check_untyped_defs = true
 
 [build-system]
-requires = ["poetry>=0.12"]
+requires = ["poetry>=1.4"]
```
Comment (alansaul) [inline pyproject.toml:96]: Note sure what this does? doesn't seem to be used for the github actions, but thought I should keep it in sync anyway?

File: .github/workflows/quality-checks.yaml (hunk 9 above)
Comment (hstojic) [reply .github/workflows/quality-checks.yaml:30]: that's fine

File: pyproject.toml (hunk 13 above)
Comment (hstojic) [reply pyproject.toml:96]: not sure either, but yes, better to keep it in sync

File: scripts/run_experiment.py [hunk 14]
```diff
@@ -106,22 +106,19 @@
     # create problem, model and acquisition rule builders based on the factories
```
Comment (hstojic) [inline scripts/run_experiment.py:106]: do we need this script if the hydra one works without issues?

Comment (hstojic) [reply scripts/run_experiment.py:106]: perhaps as a debug-experiment script?

File: scripts/run_experiment_hydra.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+def run_experiment_hydra(
+    cfg: DictConfig,
+) -> ExperimentRunner:
+    """
+    Hydra based execution. By running this method, the configuration will automatically be filled
+    Hydra will change the directory to a new working directory specific for this run so all logs
+    and configurations will be saved there.
+
+    Example usage:
+        poetry run python3 scripts/run_experiment_hydra.py acquisition=ei \
```
Comment (hstojic) [inline scripts/run_experiment_hydra.py:38]: you shouldn't need python3, correct python version is determined with `poetry install`, no?

File: scripts/run_experiment_hydra.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        reinitialise_models=cfg.reinitialise_models,
+        num_initial_points=cfg.num_initial_points,
+        num_acquisitions=cfg.num_acquisitions,
+        seed=cfg.seed,
+        experiment_name=cfg.experiment_name,
+    )
+
+    experiment_runner.run()
+    experiment_runner.save(result_directory)
+    return experiment_runner
```
Comment (hstojic) [inline scripts/run_experiment_hydra.py:66]: why do we need to return the runner?
I guess we can but we aren't doing anything with it

File: tests/integration/test_experiment.py [hunk 17]
```diff
@@ -62,20 +102,15 @@ def _test_run_experiment_produces_expected_output(
 
     with tempfile.TemporaryDirectory() as tmpdirname:
         try:
-            summary_writer = tf.summary.create_file_writer(f"{tmpdirname}/tensorboard")
```
Comment (hstojic) [inline tests/integration/test_experiment.py:65]: why removing this check?

File: tests/integration/test_experiment.py [hunk 18]
```diff
@@ -12,19 +12,54 @@
...
+
+
+@pytest.mark.parametrize(
+    "model_override,maxiter_override",
+    [
+        ("model=gpr", "model.builder.maxiter=5"),
+        ("model=svgp", "model.builder.maxiter=5"),
+        ("model=deep_ensemble", "model.builder.epochs=2"),
+        ("model=mcdropout", "model.builder.epochs=2"),
+        ("model=deep_gp", "model.builder.epochs=2"),
```
Comment (hstojic) [inline tests/integration/test_experiment.py:50]: why not 1 instead of 5 and 2 to reduce the test time further - for MC dropout and deep ensemble also reducing number of layers and units will help as well, for svgp and dgp number of inducing points...

File: tests/integration/test_experiment.py [hunk 19]
```diff
@@ -12,19 +12,54 @@
...
+)
+def test_default_run_experiment_hydra_smoke(
+    tmp_path: Path, model_override: str, maxiter_override: str
+):
+    # Use temporary directory to avoid polluting directories
+    with new_working_directory(tmp_path):
+        default_config = "config"
+        with initialize(version_base=None, config_path="../../experiment/conf"):
+            cfg = compose(config_name=default_config, overrides=[model_override, maxiter_override])
+            experiment_runner = run_experiment_hydra(cfg)
```
Comment (hstojic) [inline tests/integration/test_experiment.py:61]: I see now why you were returning the runner...

File: experiment/experiment.py (hunk 10 above)
Comment (hstojic) [reply experiment/experiment.py:95]: ok, I'll raise an issue in the project so we don't lose track of it

File: experiment/experiment.py [hunk 20]
```diff
@@ -112,49 +110,47 @@ def _init(self) -> None:
...
+                acquisition_rule=self.acquisition_rule,
             )
 
-    def _initialise_models(self) -> Mapping[Tag, ProbabilisticModel]:
+    def _initialise_models(self, data: Mapping[Tag, Dataset]) -> Mapping[Tag, ProbabilisticModel]:
         """
         Initialises the models. It is also used if `reinitialise_models` argument is set to `True`,
         in which case models are recreated at each active learning step. For deep learning models
         typically we want to do that, reusing the parameters from previous iteration leads to worse
         performance.
```
Comment (hstojic) [inline experiment/experiment.py:131]: can you add the parameter to docs please?

File: experiment/experiment.py (hunk 11 above)
Comment (hstojic) [reply experiment/experiment.py:139]: `get_initial_data`?

**Review by hstojic — APPROVED** [review]
looks good, left some minor comments but after that that's it I think

File: experiment/experiment.py (hunk 10 above)
Comment (alansaul) [reply experiment/experiment.py:95]: :thumbs-up:

File: experiment/problems.py (hunk 12 above)
Comment (alansaul) [reply experiment/problems.py:26]: When I think of a factory, I think of something that is able to create new instances (i.e. a call to the model_builder `create` method will create a new instance), this dictionary will just return an already created instance of a problem, so it's more of a look-up table to created instances, rather than a lookup table to something that is able to create an instance (a factory method/function):

https://github.com/secondmind-labs/trieste/blob/583391eb0a058eef1726fe14ed3963b3434dfaa7/trieste/objectives/single_objectives.py#L119

File: scripts/run_experiment_hydra.py (hunk 15 above)
Comment (alansaul) [reply scripts/run_experiment_hydra.py:38]: Yeah, either is fine. I'll delete the 3.

File: tests/integration/test_experiment.py (hunk 18 above)
Comment (alansaul) [reply tests/integration/test_experiment.py:50]: Personally I like to see if it goes a couple of steps, but can reduce it to 2. Agree about the SVGP and DGP, will try and restrict these with more overrides.

File: scripts/run_experiment.py (hunk 14 above)
Comment (alansaul) [reply scripts/run_experiment.py:106]: Yeah I'm not sure. I feel like it can probably be deleted, I should make sure they have the some functionality first.

File: scripts/run_experiment_hydra.py (hunk 16 above)
Comment (alansaul) [reply scripts/run_experiment_hydra.py:66]: Yeah in general I think if we don't return this, there would be currently no way of probing the models that are trained, or anything, so I think it makes sense to return it as I think we will want access in practice.

File: tests/integration/test_experiment.py (hunk 18 above)
Comment (alansaul) [reply tests/integration/test_experiment.py]: I ended up just making some test configs that override the default expensive terms with cheaper ones. Can adjust is testing is still too slow.
